File: server/tags.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const DATA_DIR = path.join(__dirname, '../data');
const PROJECT_TAGS_FILE = path.join(DATA_DIR, 'project-tags.json');
const TAG_DEFINITIONS_FILE = path.join(DATA_DIR, 'tag-definitions.json');
const TAG_COLORS_FILE = path.join(DATA_DIR, 'tag-colors.json');

// 진행 관련 태그 (시스템 고정)
const PROGRESS_TAGS = ['진행중', '중지', '완료', '계획중', 'deprecated'];

// 초기 구분 태그
const DEFAULT_CATEGORY_TAGS = [
  'AI', '웹앱', 'CLI', '봇', '스크래퍼', '자동화',
  '지식관리', 'Obsidian', '에이전트', '데이터수집', '문서변환'
];

// Tailwind 색상 팔레트
const TAILWIND_COLORS = [
  'bg-purple-100 text-purple-800',
  'bg-indigo-100 text-indigo-800',
  'bg-blue-100 text-blue-800',
  'bg-cyan-100 text-cyan-800',
  'bg-teal-100 text-teal-800',
  'bg-emerald-100 text-emerald-800',
  'bg-green-100 text-green-800',
  'bg-lime-100 text-lime-800',
  'bg-yellow-100 text-yellow-800',
  'bg-amber-100 text-amber-800',
  'bg-orange-100 text-orange-800',
  'bg-pink-100 text-pink-800',
  'bg-rose-100 text-rose-800',
  'bg-violet-100 text-violet-800',
  'bg-fuchsia-100 text-fuchsia-800'
];

/**
 * 데이터 디렉토리 초기화
 */
function ensureDataDir() {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
}

/**
 * 프로젝트 태그 데이터 로드
 */
export function loadProjectTags() {
  ensureDataDir();

  if (!fs.existsSync(PROJECT_TAGS_FILE)) {
    return {};
  }

  try {
    const data = fs.readFileSync(PROJECT_TAGS_FILE, 'utf-8');
    return JSON.parse(data);
  } catch (error) {
    console.error('Error loading project tags:', error);
    return {};
  }
}

/**
 * 프로젝트 태그 데이터 저장
 */
export function saveProjectTags(tags) {
  ensureDataDir();

  try {
    fs.writeFileSync(PROJECT_TAGS_FILE, JSON.stringify(tags, null, 2), 'utf-8');
    return true;
  } catch (error) {
    console.error('Error saving project tags:', error);
    return false;
  }
}

/**
 * 태그 정의 로드
 */
export function loadTagDefinitions() {
  ensureDataDir();

  if (!fs.existsSync(TAG_DEFINITIONS_FILE)) {
    const defaults = {
      progress: PROGRESS_TAGS,
      categories: DEFAULT_CATEGORY_TAGS
    };
    saveTagDefinitions(defaults);
    return defaults;
  }

  try {
    const data = fs.readFileSync(TAG_DEFINITIONS_FILE, 'utf-8');
    return JSON.parse(data);
  } catch (error) {
    console.error('Error loading tag definitions:', error);
    return {
      progress: PROGRESS_TAGS,
      categories: DEFAULT_CATEGORY_TAGS
    };
  }
}

/**
 * 태그 정의 저장
 */
export function saveTagDefinitions(definitions) {
  ensureDataDir();

  try {
    fs.writeFileSync(TAG_DEFINITIONS_FILE, JSON.stringify(definitions, null, 2), 'utf-8');
    return true;
  } catch (error) {
    console.error('Error saving tag definitions:', error);
    return false;
  }
}

/**
 * 태그 색상 로드
 */
export function loadTagColors() {
  ensureDataDir();

  if (!fs.existsSync(TAG_COLORS_FILE)) {
    const defaults = {
      progress: {
        '진행중': 'bg-green-100 text-green-800 border-green-300',
        '중지': 'bg-gray-100 text-gray-800 border-gray-300',
        '완료': 'bg-blue-100 text-blue-800 border-blue-300',
        '계획중': 'bg-orange-100 text-orange-800 border-orange-300',
        'deprecated': 'bg-red-100 text-red-800 border-red-300'
      },
      categories: {
        'AI': 'bg-purple-100 text-purple-800',
        '웹앱': 'bg-indigo-100 text-indigo-800',
        'CLI': 'bg-slate-100 text-slate-800',
        '봇': 'bg-pink-100 text-pink-800',
        '스크래퍼': 'bg-orange-100 text-orange-800',
        '자동화': 'bg-cyan-100 text-cyan-800',
        '지식관리': 'bg-amber-100 text-amber-800',
        'Obsidian': 'bg-violet-100 text-violet-800',
        '에이전트': 'bg-fuchsia-100 text-fuchsia-800',
        '데이터수집': 'bg-teal-100 text-teal-800',
        '문서변환': 'bg-emerald-100 text-emerald-800',
        'default': 'bg-gray-100 text-gray-600'
      }
    };
    saveTagColors(defaults);
    return defaults;
  }

  try {
    const data = fs.readFileSync(TAG_COLORS_FILE, 'utf-8');
    return JSON.parse(data);
  } catch (error) {
    console.error('Error loading tag colors:', error);
    return {};
  }
}

/**
 * 태그 색상 저장
 */
export function saveTagColors(colors) {
  ensureDataDir();

  try {
    fs.writeFileSync(TAG_COLORS_FILE, JSON.stringify(colors, null, 2), 'utf-8');
    return true;
  } catch (error) {
    console.error('Error saving tag colors:', error);
    return false;
  }
}

/**
 * 랜덤 색상 할당
 */
export function assignRandomColor() {
  return TAILWIND_COLORS[Math.floor(Math.random() * TAILWIND_COLORS.length)];
}

/**
 * 구분 태그 추가
 */
export function addCategoryTag(tag) {
  // 디버그 로그 파일에 기록
  fs.appendFileSync('/tmp/tag-debug.log', `\n[${new Date().toISOString()}] addCategoryTag 호출: ${tag}\n`);

  const definitions = loadTagDefinitions();
  fs.appendFileSync('/tmp/tag-debug.log', `definitions: ${JSON.stringify(definitions.categories)}\n`);

  if (definitions.categories.includes(tag)) {
    fs.appendFileSync('/tmp/tag-debug.log', '이미 존재하는 태그\n');
    return { success: false, message: 'Tag already exists' };
  }

  definitions.categories.push(tag);
  fs.appendFileSync('/tmp/tag-debug.log', `태그 추가 후: ${JSON.stringify(definitions.categories)}\n`);

  const saveResult = saveTagDefinitions(definitions);
  fs.appendFileSync('/tmp/tag-debug.log', `저장 결과: ${saveResult}\n`);

  // 새 태그에 색상 할당
  const colors = loadTagColors();
  if (!colors.categories[tag]) {
    colors.categories[tag] = assignRandomColor();
    saveTagColors(colors);
  }

  const result = { success: true, definitions, colors };
  fs.appendFileSync('/tmp/tag-debug.log', `반환값: ${JSON.stringify(result).substring(0, 200)}\n`);

  return result;
}

/**
 * 구분 태그 삭제
 */
export function deleteCategoryTag(tag) {
  const definitions = loadTagDefinitions();

  // 태그가 존재하는지 확인
  const index = definitions.categories.indexOf(tag);
  if (index === -1) {
    return { success: false, message: 'Tag not found' };
  }

  // 태그를 사용 중인 프로젝트 확인
  const projectTags = loadProjectTags();
  const projectsUsingTag = Object.entries(projectTags)
    .filter(([_, data]) => data.categories && data.categories.includes(tag))
    .map(([name, _]) => name);

  if (projectsUsingTag.length > 0) {
    // 경고와 함께 사용 중인 프로젝트 목록 반환
    return {
      success: false,
      message: 'Tag is in use',
      projectsUsingTag
    };
  }

  // 태그 삭제
  definitions.categories.splice(index, 1);
  saveTagDefinitions(definitions);

  // 색상 정의도 삭제
  const colors = loadTagColors();
  delete colors.categories[tag];
  saveTagColors(colors);

  return { success: true, definitions, colors };
}

/**
 * 태그 색상 업데이트
 */
export function updateTagColor(tag, color) {
  const colors = loadTagColors();

  colors.categories[tag] = color;
  saveTagColors(colors);

  return { success: true, colors };
}

/**
 * 특정 프로젝트의 태그 가져오기
 */
export function getProjectTags(projectName) {
  const allTags = loadProjectTags();
  return allTags[projectName] || {
    customTitle: null,
    progress: '계획중',
    categories: [],
    favorite: false,
    archived: false,
    notes: ''
  };
}

/**
 * 특정 프로젝트의 태그 저장
 */
export function setProjectTags(projectName, tags) {
  const allTags = loadProjectTags();

  // customTitle 유효성 검사
  if (tags.customTitle !== undefined) {
    const trimmed = tags.customTitle?.trim();
    tags.customTitle = trimmed && trimmed.length <= 50 ? trimmed : null;
  }

  allTags[projectName] = tags;
  saveProjectTags(allTags);

  return { success: true };
}

File: src-tauri/src/tags.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ProjectTags {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub custom_title: Option<String>,
    pub progress: String,
    pub categories: Vec<String>,
    pub favorite: bool,
    pub archived: bool,
    pub notes: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TagDefinitions {
    pub progress: Vec<String>,
    pub categories: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TagColors {
    pub progress: HashMap<String, String>,
    pub categories: HashMap<String, String>,
}

const PROGRESS_TAGS: &[&str] = &["진행중", "중지", "완료", "계획중", "deprecated"];

const DEFAULT_CATEGORY_TAGS: &[&str] = &[
    "AI",
    "웹앱",
    "CLI",
    "봇",
    "스크래퍼",
    "자동화",
    "지식관리",
    "Obsidian",
    "에이전트",
    "데이터수집",
    "문서변환",
];

const TAILWIND_COLORS: &[&str] = &[
    "bg-purple-100 text-purple-800",
    "bg-indigo-100 text-indigo-800",
    "bg-blue-100 text-blue-800",
    "bg-cyan-100 text-cyan-800",
    "bg-teal-100 text-teal-800",
    "bg-emerald-100 text-emerald-800",
    "bg-green-100 text-green-800",
    "bg-lime-100 text-lime-800",
    "bg-yellow-100 text-yellow-800",
    "bg-amber-100 text-amber-800",
    "bg-orange-100 text-orange-800",
    "bg-pink-100 text-pink-800",
    "bg-rose-100 text-rose-800",
    "bg-violet-100 text-violet-800",
    "bg-fuchsia-100 text-fuchsia-800",
];

fn get_data_dir() -> PathBuf {
    // 프로젝트 루트의 data 디렉토리 사용
    let data_dir = PathBuf::from("data");
    if !data_dir.exists() {
        fs::create_dir_all(&data_dir).ok();
    }
    data_dir
}

fn get_project_tags_file() -> PathBuf {
    get_data_dir().join("project-tags.json")
}

fn get_tag_definitions_file() -> PathBuf {
    get_data_dir().join("tag-definitions.json")
}

fn get_tag_colors_file() -> PathBuf {
    get_data_dir().join("tag-colors.json")
}

/// 프로젝트 태그 데이터 로드
pub fn load_project_tags() -> HashMap<String, ProjectTags> {
    let file_path = get_project_tags_file();

    if !file_path.exists() {
        return HashMap::new();
    }

    match fs::read_to_string(&file_path) {
        Ok(content) => serde_json::from_str(&content).unwrap_or_default(),
        Err(_) => HashMap::new(),
    }
}

/// 프로젝트 태그 데이터 저장
pub fn save_project_tags(tags: &HashMap<String, ProjectTags>) -> Result<(), String> {
    let file_path = get_project_tags_file();

    let content = serde_json::to_string_pretty(tags)
        .map_err(|e| format!("Failed to serialize tags: {}", e))?;

    fs::write(&file_path, content).map_err(|e| format!("Failed to write tags file: {}", e))?;

    Ok(())
}

/// 태그 정의 로드
pub fn load_tag_definitions() -> TagDefinitions {
    let file_path = get_tag_definitions_file();

    if !file_path.exists() {
        let defaults = TagDefinitions {
            progress: PROGRESS_TAGS.iter().map(|s| s.to_string()).collect(),
            categories: DEFAULT_CATEGORY_TAGS.iter().map(|s| s.to_string()).collect(),
        };
        save_tag_definitions(&defaults).ok();
        return defaults;
    }

    match fs::read_to_string(&file_path) {
        Ok(content) => serde_json::from_str(&content).unwrap_or_else(|_| TagDefinitions {
            progress: PROGRESS_TAGS.iter().map(|s| s.to_string()).collect(),
            categories: DEFAULT_CATEGORY_TAGS.iter().map(|s| s.to_string()).collect(),
        }),
        Err(_) => TagDefinitions {
            progress: PROGRESS_TAGS.iter().map(|s| s.to_string()).collect(),
            categories: DEFAULT_CATEGORY_TAGS.iter().map(|s| s.to_string()).collect(),
        },
    }
}

/// 태그 정의 저장
pub fn save_tag_definitions(definitions: &TagDefinitions) -> Result<(), String> {
    let file_path = get_tag_definitions_file();

    let content = serde_json::to_string_pretty(definitions)
        .map_err(|e| format!("Failed to serialize definitions: {}", e))?;

    fs::write(&file_path, content)
        .map_err(|e| format!("Failed to write definitions file: {}", e))?;

    Ok(())
}

/// 태그 색상 로드
pub fn load_tag_colors() -> TagColors {
    let file_path = get_tag_colors_file();

    if !file_path.exists() {
        let defaults = get_default_tag_colors();
        save_tag_colors(&defaults).ok();
        return defaults;
    }

    match fs::read_to_string(&file_path) {
        Ok(content) => serde_json::from_str(&content).unwrap_or_else(|_| get_default_tag_colors()),
        Err(_) => get_default_tag_colors(),
    }
}

fn get_default_tag_colors() -> TagColors {
    let mut progress = HashMap::new();
    progress.insert(
        "진행중".to_string(),
        "bg-green-100 text-green-800 border-green-300".to_string(),
    );
    progress.insert(
        "중지".to_string(),
        "bg-gray-100 text-gray-800 border-gray-300".to_string(),
    );
    progress.insert(
        "완료".to_string(),
        "bg-blue-100 text-blue-800 border-blue-300".to_string(),
    );
    progress.insert(
        "계획중".to_string(),
        "bg-orange-100 text-orange-800 border-orange-300".to_string(),
    );
    progress.insert(
        "deprecated".to_string(),
        "bg-red-100 text-red-800 border-red-300".to_string(),
    );

    let mut categories = HashMap::new();
    categories.insert("AI".to_string(), "bg-purple-100 text-purple-800".to_string());
    categories.insert("웹앱".to_string(), "bg-indigo-100 text-indigo-800".to_string());
    categories.insert("CLI".to_string(), "bg-slate-100 text-slate-800".to_string());
    categories.insert("봇".to_string(), "bg-pink-100 text-pink-800".to_string());
    categories.insert(
        "스크래퍼".to_string(),
        "bg-orange-100 text-orange-800".to_string(),
    );
    categories.insert("자동화".to_string(), "bg-cyan-100 text-cyan-800".to_string());
    categories.insert(
        "지식관리".to_string(),
        "bg-amber-100 text-amber-800".to_string(),
    );
    categories.insert(
        "Obsidian".to_string(),
        "bg-violet-100 text-violet-800".to_string(),
    );
    categories.insert(
        "에이전트".to_string(),
        "bg-fuchsia-100 text-fuchsia-800".to_string(),
    );
    categories.insert(
        "데이터수집".to_string(),
        "bg-teal-100 text-teal-800".to_string(),
    );
    categories.insert(
        "문서변환".to_string(),
        "bg-emerald-100 text-emerald-800".to_string(),
    );
    categories.insert("default".to_string(), "bg-gray-100 text-gray-600".to_string());

    TagColors {
        progress,
        categories,
    }
}

/// 태그 색상 저장
pub fn save_tag_colors(colors: &TagColors) -> Result<(), String> {
    let file_path = get_tag_colors_file();

    let content = serde_json::to_string_pretty(colors)
        .map_err(|e| format!("Failed to serialize colors: {}", e))?;

    fs::write(&file_path, content).map_err(|e| format!("Failed to write colors file: {}", e))?;

    Ok(())
}

/// 랜덤 색상 할당
pub fn assign_random_color() -> String {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let index = rng.gen_range(0..TAILWIND_COLORS.len());
    TAILWIND_COLORS[index].to_string()
}

/// 구분 태그 추가
pub fn add_category_tag(tag: &str) -> Result<(TagDefinitions, TagColors), String> {
    let mut definitions = load_tag_definitions();

    if definitions.categories.contains(&tag.to_string()) {
        return Err("Tag already exists".to_string());
    }

    definitions.categories.push(tag.to_string());
    save_tag_definitions(&definitions)?;

    // 새 태그에 색상 할당
    let mut colors = load_tag_colors();
    if !colors.categories.contains_key(tag) {
        colors
            .categories
            .insert(tag.to_string(), assign_random_color());
        save_tag_colors(&colors)?;
    }

    Ok((definitions, colors))
}

/// 구분 태그 삭제
pub fn delete_category_tag(tag: &str) -> Result<(TagDefinitions, TagColors), String> {
    let mut definitions = load_tag_definitions();

    let index = definitions
        .categories
        .iter()
        .position(|t| t == tag)
        .ok_or("Tag not found".to_string())?;

    // 태그를 사용 중인 프로젝트 확인
    let project_tags = load_project_tags();
    let projects_using_tag: Vec<String> = project_tags
        .iter()
        .filter(|(_, tags)| tags.categories.contains(&tag.to_string()))
        .map(|(name, _)| name.clone())
        .collect();

    if !projects_using_tag.is_empty() {
        return Err(format!("Tag is in use by: {:?}", projects_using_tag));
    }

    // 태그 삭제
    definitions.categories.remove(index);
    save_tag_definitions(&definitions)?;

    // 색상 정의도 삭제
    let mut colors = load_tag_colors();
    colors.categories.remove(tag);
    save_tag_colors(&colors)?;

    Ok((definitions, colors))
}

/// 태그 색상 업데이트
pub fn update_tag_color(tag: &str, color: &str) -> Result<TagColors, String> {
    let mut colors = load_tag_colors();
    colors.categories.insert(tag.to_string(), color.to_string());
    save_tag_colors(&colors)?;
    Ok(colors)
}

/// 특정 프로젝트의 태그 가져오기
pub fn get_project_tags(project_name: &str) -> ProjectTags {
    let all_tags = load_project_tags();

    all_tags.get(project_name).cloned().unwrap_or(ProjectTags {
        custom_title: None,
        progress: "계획중".to_string(),
        categories: Vec::new(),
        favorite: false,
        archived: false,
        notes: String::new(),
    })
}

/// 특정 프로젝트의 태그 저장
pub fn set_project_tags(project_name: &str, mut tags: ProjectTags) -> Result<(), String> {
    // customTitle 유효성 검사
    if let Some(ref title) = tags.custom_title {
        let trimmed = title.trim();
        if trimmed.is_empty() || trimmed.len() > 50 {
            tags.custom_title = None;
        } else {
            tags.custom_title = Some(trimmed.to_string());
        }
    }

    let mut all_tags = load_project_tags();
    all_tags.insert(project_name.to_string(), tags);
    save_project_tags(&all_tags)?;

    Ok(())
}
